Add backblaze b2 backup integration

ElCruncharino opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
Adds a new core integration for Backblaze B2. Can be used as a backup source for HA Backups.

This PR builds upon the foundational work in #145508 by @hugo-vrijswijk  and incorporates feedback received on that PR.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

As @hugo-vrijswijk mentioned in their PR, `b2sdk` is completely sync, so that needed some refactoring calls wrapped in `async_add_executor_job`.

Primary changes (based on ongoing discussions in that PR):
- **Separate Metadata Files**: Backups now consist of two files in Backblaze B2: the main backup archive (.tar) and a separate JSON metadata file (.metadata.json).
- **Optimized Uploads**: Implemented distinct upload strategies for small (direct upload) and large (multipart upload via temporary file) backups, enhancing efficiency and stability. I wanted to avoid keeping uploads in memory.
- **Caching:** Introduced caching for the list of backups and file information, mostly to reduce API calls to Backblaze B2 and improving responsiveness in instances with lots of backups.
- **Backup Lifecycle Management**: Improved logic for listing, retrieving, and deleting backups by leveraging the new metadata file structure.
- **Stream Handling**: Added more explicit closing of file and network streams during downloads and uploads to prevent resource leaks.

Tests were overhauled to get full coverage, but I'm relatively new to writing tests for these types of integrations, so that was a lot of trial and error.

I've also tested locally in the UI with large backups, small backups, and so on with a real B2 bucket, and concluded there were no major functional or performance issues with the code.

Much like @hugo-vrijswijk, this is my first time contributing to HA, so if 

- ~~This PR fixes or closes issue: fixes #~~
- This PR is related to issue: #144497, #143995, #144474
- Link to documentation pull request: @hugo-vrijswijk's documentation is compatible home-assistant/home-assistant.io#39169
- Link to brands pull request: home-assistant/brands#7125
- ~~Link to developer documentation pull request:~~
- ~~Link to frontend pull request:~~ 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
    - @hugo-vrijswijk has documentation in his pull request, home-assistant/home-assistant.io#39169. It should largely be compatible. I only added and refactored internal logic.

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

hugo-vrijswijk committed:
Add Backblaze B2 integration

hugo-vrijswijk committed:
Add tests for backup and init

hugo-vrijswijk committed:
Improve test coverage

hugo-vrijswijk committed:
Update log-when-unavailable quality_scale

hugo-vrijswijk committed:
Rename backblaze_b2 to backblaze

hugo-vrijswijk committed:
Wrap in single executor job

hugo-vrijswijk committed:
Extract AgentBackup creation to method

hugo-vrijswijk committed:
Remove info log

hugo-vrijswijk committed:
Automatically add prefix slash

hugo-vrijswijk committed:
Remove cast

hugo-vrijswijk committed:
Move some code outside try-catch

hugo-vrijswijk committed:
Update translation string for prefix directory path

hugo-vrijswijk committed:
Use unique entry_id

hugo-vrijswijk committed:
Update imports and assert step_id and errors

hugo-vrijswijk committed:
Fix test

ElCruncharino committed:
Working implementation of file-based metadata. Tests passing.

ElCruncharino committed:
Refactoring and improvements

ElCruncharino committed:
Increase test coverage to 97%

ElCruncharino committed:
Updated to 98% coverage in tests.

ElCruncharino committed:
Updated to fix upload/download

ElCruncharino committed:
Allow for multipart uploads

ElCruncharino committed:
Fixed tests

ElCruncharino committed:
Updates for downloading backups, and some test changes.

ElCruncharino committed:
Updated some tests and improved upload sequence

ElCruncharino committed:
Improved tests and cleaned up comments

ElCruncharino committed:
Continued updates to tests. Minor tweaks to backup for multipart uploads.

ElCruncharino committed:
Improved coverage

ElCruncharino committed:
98% coverage

ElCruncharino committed:
100% coverage

ElCruncharino committed:
Refactoring, and simpler tests

ElCruncharino committed:
Revert to large test suite with 100% coverage.

ElCruncharino committed:
Added caching

ElCruncharino committed:
Refactoring of some tests and config flow.

ElCruncharino committed:
enhance backblaze integration to gold quality scale

- add reauthentication and reconfiguration flows
- implement diagnostics and repair issues for troubleshooting
- improve cache invalidation and code organization
- add comprehensive tests and strict typing
- update quality scale: bronze → silver → gold

ElCruncharino committed:
logging import for config_flow

ElCruncharino committed:
consolidate backblaze tests, reduce by 58% while keeping 100% coverage

ElCruncharino committed:
improve backblaze test coverage from 84% to 91%

ElCruncharino committed:
add safety limits and validation to backblaze component

ElCruncharino committed:
improve backblaze test coverage to 94%

ElCruncharino committed:
add targeted tests to improve coverage to 93%

ElCruncharino committed:
add targeted tests for 94% coverage

ElCruncharino committed:
reach 94% coverage with line 583 cache scenario

ElCruncharino committed:
refactor backblaze test suite

ElCruncharino committed:
fixed linting

ElCruncharino committed:
updated JSON metadata formatting

emontnemery reviewed (commented):
A few comments. I want to understand how the file iteration is supposed to work before I continue the review.

ElCruncharino committed:
address emontnemery review feedback for backblaze integration

ElCruncharino commented on homeassistant/components/backblaze/__init__.py:
```diff
@@ -0,0 +1,135 @@
+        raise ConfigEntryError(
+            translation_domain=DOMAIN,
+            translation_key="invalid_credentials",
+        ) from err
```
Ah, you're totally right about this. I completely missed that ConfigEntryAuthFailed is what triggers the reauth flow. I was just throwing ConfigEntryError for everything auth-related. Fixed it to use ConfigEntryAuthFailed for auth failures so users actually get the reauth prompt instead of just a broken integration.

ElCruncharino commented on homeassistant/components/backblaze/strings.json:
```diff
@@ -0,0 +1,122 @@
+    "bucket_not_found": {
+      "title": "Backblaze bucket not found",
+      "description": "The Backblaze bucket '{bucket_name}' cannot be found or accessed. This could mean:\n\n1. The bucket was deleted\n2. The bucket name was changed\n3. Your credentials no longer have access to this bucket\n\nTo fix this issue:\n\n1. Log in to your Backblaze account\n2. Verify the bucket still exists and check its name\n3. Ensure your application key has access to this bucket\n4 ...
+    }
```
I was just sloppy with the repair flow descriptions. I was copy-pasting text that suggested clicking Submit would actually fix things, when really it just dismisses the issue. Updated the descriptions to indicate that Submit dismisses the issue and users need to manually reconfigure through the settings

ElCruncharino commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,819 @@
+            vol.Optional("extra_metadata", default=dict): dict,
+        },
+    }
+)
```
No worries. Thanks for pointing that out. I'm new to the backup system architecture here. Removed that.

ElCruncharino commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,819 @@
+    async def _close_response(self, response: Any) -> None:
+        """Close a response object gracefully."""
+        if hasattr(response, "close") and callable(response.close):
+            await self._hass.async_add_executor_job(response.close)
```
Yeah, that was pretty messy. I had some unclear typing around stream objects and response handling because I was struggling with a few of the tests I was working with. Cleaned up the type annotations and made the object handling more explicit so it's clearer what each method expects.

ElCruncharino commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,819 @@
+            _LOGGER.exception(
+                "Manual intervention may be required to delete %s from Backblaze B2",
+                filename,
+            )
```
I ended up with redundant logging in a couple places. I remember putting this in there for some testing clarity when I was reading through the caplog stuff. Cleaned that up to avoid the noise in the logs

ElCruncharino commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,819 @@
+            try:
+                # B2 SDK's response.iter_content is blocking, so stream it chunk by chunk
+                # within the executor.
+                iterator = response.iter_content(chunk_size=8192)
```
Absolutely! 8KB was too small and I would imagine causing unnecessary thread switching overhead. Bumped it up to 1MB chunks for much better performance, especially for larger backups.

ElCruncharino commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,819 @@
+                "backup_id": backup.backup_id,
+                "backup_metadata": backup.as_dict(),
+            },
+            indent=2,
```
A user can, since its in the bucket. I based this on the S3 backup integration (which frankly I used a lot of as a basis here), but even that doesn't pretty print, so it's just wasted bytes. I did like the pretty print for when I was visually inspecting the JSON in the bucket, but its not necessary. Removed the indent=2 to keep things minimal.

ElCruncharino committed:
fix: use correct B2 SDK upload_unbound_stream method

emontnemery commented:
@ElCruncharino if you want to discuss something, feel free to message me on Discord

emontnemery commented on homeassistant/components/backblaze/__init__.py:
```diff
@@ -63,7 +68,22 @@ def _authorize_and_get_bucket_sync() -> Bucket:
             err,
         )
-        raise ConfigEntryError(
+        # Create repair issue for user to address
+        ir.async_create_issue(
+            hass,
+            DOMAIN,
+            f"bucket_access_restricted_{entry.entry_id}",
+            is_fixable=True,
+            issue_domain=DOMAIN,
+            severity=ir.IssueSeverity.ERROR,
+            translation_key="bucket_access_restricted",
+            translation_placeholders={
+                "title": entry.title,
+                "bucket_name": err.bucket_name,
+                "entry_id": entry.entry_id,
+            },
+        )
```
Maybe add helpers in `repairs.py` for creating the repairs to avoid duplicating it?

emontnemery commented on homeassistant/components/backblaze/__init__.py:
```diff
@@ -77,7 +97,22 @@ def _authorize_and_get_bucket_sync() -> Bucket:
             err,
         )
-        raise ConfigEntryError(
+        # Create repair issue for user to address
+        ir.async_create_issue(
+            hass,
+            DOMAIN,
+            f"bucket_not_found_{entry.entry_id}",
+            is_fixable=True,
+            issue_domain=DOMAIN,
+            severity=ir.IssueSeverity.ERROR,
+            translation_key="bucket_not_found",
+            translation_placeholders={
+                "title": entry.title,
+                "bucket_name": entry.data.get(CONF_BUCKET, "unknown"),
+                "entry_id": entry.entry_id,
+            },
+        )
```
Same comment

ElCruncharino replied:
Same as above.

emontnemery commented on homeassistant/components/backblaze/strings.json:
```diff
@@ -80,43 +80,13 @@
     "bucket_access_restricted": {
       "title": "Backblaze bucket access restricted",
-      "description": "Access to your Backblaze bucket '{bucket_name}' is restricted for the current credentials. This means your application key may only have access to specific buckets, but not this one. To fix this issue:\n\n1. Log in to your Backblaze account\n2. Check your application key restrictions\n3. Either use a different bucket that your key can access, or create a new application key  ...
+      "description": "Access to your Backblaze bucket {bucket_name} is restricted for the current credentials. This means your application key may only have access to specific buckets, but not this one. To fix this issue:\n\n1. Log in to your Backblaze account\n2. Check your application key restrictions\n3. Either use a different bucket that your key can access, or create a new application key wi ...
```
Instead of telling the user to click submit, tell them that the issue can be fixed by completing the reconfigure flow.
Finishing the reconfigure flow should reload the config entry, and when that happens you should clear up issues.

emontnemery commented on homeassistant/components/backblaze/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "domain": "backblaze",
+  "name": "Backblaze",
```
Just a control question, would it be more suitable to name the integration "Backblaze B2", or is it enough with just "Backblaze"?

ElCruncharino replied:
I moved the naming scheme to backblaze_b2. This followed some other backup components with the company_product schema for the name: aws_s3, azure_storage, google_drive, etc. @hugo-vrijswijk You'll need to change the branding pull request names if you don't mind.

emontnemery commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,678 @@
+        finally:
+            # Close the reader to clean up resources
+            reader.close()
+            await self._close_stream(stream)
```
I don't think we need to call `aclose` on the object returned by the `open_stream` function. Why did you add it? Did you see resource leakage if you don't call it?

emontnemery commented on homeassistant/components/backblaze/backup.py:
```diff
@@ -0,0 +1,678 @@
+                    yield chunk
+            finally:
+                _LOGGER.debug("Closing download stream for %s", file.file_name)
+                # Ensure the underlying response stream is closed
+                await self._close_requests_response(response)
```
Why do we need to call close on the response?

According to the `requests.close` docstring, it should not be necessary.

ElCruncharino replied:
Same as above. Old habit from file handles and connections. I was treating the response object like a file descriptor that needed explicit cleanup, but Python's requests library handles this automatically, as you pointed out.

emontnemery reviewed (commented):
Some follow-up questions and comments.

ElCruncharino committed:
address emontnemery followup review feedback for backblaze integration

ElCruncharino commented on homeassistant/components/backblaze/__init__.py:
```diff
@@ -0,0 +1,170 @@
+        raise ConfigEntryError(
+            translation_domain=DOMAIN,
+            translation_key="invalid_auth",
+        ) from err
```
I feel dumb for not vetting all of the reauth flow edges cases, but I'm starting to get it. Thanks to your patience. Changed **MissingAccountData** to raise **ConfigEntryAuthFailed** instead of **ConfigEntryError** so users get the reauth flow instead of a dead integration.

ElCruncharino committed:
Added a couple of helper functions, and reconfigured tests.

ElCruncharino committed:
inject "Backblaze" via description_placeholders & rename component to backblaze_b2

ElCruncharino commented:
> @ElCruncharino please confirm the domain you want to use for this integration, then mark the PR ready for review.

Done! "backblaze_b2", for reasons stated in the resolved conversation.

jeroenleenarts commented:
Tried finding another way to reach out.

Just wanted to say I am impressed by your perseverance in bringing this integration to completion.

@ElCruncharino Thank you. Looking forward to trying your work once completed.

emontnemery commented on homeassistant/components/backblaze_b2/diagnostics.py:
```diff
@@ -0,0 +1,68 @@
+    """Return diagnostics for a config entry."""
+    bucket = entry.runtime_data
+
+    # Get bucket information (safely)
+    try:
+        bucket_info = {
+            "name": bucket.name,
+            "id": bucket.id_,
+            "type": getattr(bucket, "type_", "unknown"),
+            "cors_rules": getattr(bucket, "cors_rules", "not_available"),
+            "lifecycle_rules": getattr(bucket, "lifecycle_rules", "not_available"),
+            "revision": getattr(bucket, "revision", "not_available"),
+        }
+    except (AttributeError, TypeError, ValueError):
+        bucket_info = {"error": "Failed to retrieve bucket information"}
```
I guess the "safely" refers to using getattr. However, why do we do that, why do we need to guard against missing attributes on the `Bucket` instance?

ElCruncharino replied:
IIRC, I did this to pass a test when I was monkeying with the diagnostics (with a Mock Bucket). Not necessary now. I updated my BucketSimulator and the tests.

emontnemery commented on homeassistant/components/backblaze_b2/diagnostics.py:
```diff
@@ -0,0 +1,68 @@
+    except (AttributeError, TypeError, ValueError):
+        bucket_info = {"error": "Failed to retrieve bucket information"}
+
+    # Get account information (safely)
```
Same question

emontnemery commented on homeassistant/components/backblaze_b2/config_flow.py:
```diff
@@ -0,0 +1,294 @@
+        assert self.reauth_entry is not None
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
```
The user flow guards against other config entries with same key id +application key, I think we need to do that here too.

ElCruncharino replied:
Sounds good. Consolidated this fix to _abort_if_duplicate_credentials to avoid some minor code duplication too.

emontnemery commented on homeassistant/components/backblaze_b2/config_flow.py:
```diff
@@ -0,0 +1,294 @@
+            },
+        )
+
+    async def async_step_reconfigure(
```
Same comment as for the reauth flow.

ElCruncharino committed:
address  emontnemery's review points in pull request

ElCruncharino commented on homeassistant/components/backblaze_b2/strings.json:
```diff
@@ -0,0 +1,92 @@
+  },
+  "issues": {
+    "bucket_access_restricted": {
+      "title": "{brand_name} bucket access restricted",
+      "description": "Access to your {brand_name} bucket {bucket_name} is restricted for the current credentials. This means your application key may only have access to specific buckets, but not this one. To fix this issue:\n\n1. Log in to your {brand_name} account\n2. Check your application key restrictions\n3. Either use a different bucket that your key can access, or create a new application  ...
+    },
+    "bucket_not_found": {
+      "title": "{brand_name} bucket not found",
+      "description": "The {brand_name} bucket {bucket_name} cannot be found or accessed. This could mean:\n\n1. The bucket was deleted\n2. The bucket name was changed\n3. Your credentials no longer have access to this bucket\n\nTo fix this issue:\n\n1. Log in to your {brand_name} account\n2. Verify the bucket still exists and check its name\n3. Ensure your application key has access to this bucke ...
+    }
```
Yep! issues should now be cleared in `async_check_for_repair_issues` when conditions are resolved. Will remove is_fixable=True since issues auto-clear when no problems detected (though mypy whines if I don't have is_fixable at all, so I guess I'll set it to False). The repair flow will guide the user to fix anyway.

ElCruncharino commented on homeassistant/components/backblaze_b2/repairs.py:
```diff
@@ -0,0 +1,122 @@
+        hass,
+        DOMAIN,
+        f"{ISSUE_BUCKET_NOT_FOUND}_{entry.entry_id}",
+        is_fixable=True,
```
Same as above.

ElCruncharino committed:
Update homeassistant/components/backblaze_b2/__init__.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
Update 'Backblaze' references to 'Backblaze B2' to match new component name.

emontnemery commented on homeassistant/components/backblaze_b2/__init__.py:
```diff
@@ -0,0 +1,150 @@
+
+    async def _periodic_issue_check(_now: Any) -> None:
+        """Periodically check for repair issues."""
+        del _now  # Required by async_track_time_interval interface
```
Please remove all the `del` statements added in yesterday's commit, our linters don't force you to reference passed in parameters.

ElCruncharino committed:
Removed dels from incorrect linting config. Cleaned up some tests.

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,614 @@
+"""Backup platform for the Backblaze B2 integration."""
+
+# pylint: disable=hass-component-root-import
+
```
Let's make this more targeted.
```suggestion
```

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,614 @@
+    BackupNotFound,
+    suggested_filename,
+)
+from homeassistant.components.backup.util import AsyncIteratorReader
```
```suggestion

# pylint: disable-next=hass-component-root-import
from homeassistant.components.backup.util import AsyncIteratorReader
```

Note: I think we should either export `AsyncIteratorReader` this from the backup integration, or move it to some async util. I'll discuss it with others and open a PR doing one of those alternatives.

emontnemery commented on homeassistant/components/backblaze_b2/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+            b2_api.get_bucket_by_name(user_input[CONF_BUCKET])
+
+        try:
+            await self.hass.async_add_executor_job(_authorize_and_get_bucket_sync)
+
+            allowed = b2_api.account_info.get_allowed()
+
+            if (
+                allowed is None
+                or not allowed.get("capabilities")
+                or not REQUIRED_CAPABILITIES.issubset(set(allowed["capabilities"]))
+            ):
+                # Check which required capabilities are missing
+                missing_caps = REQUIRED_CAPABILITIES - set(
+                    allowed.get("capabilities", [])
+                )
+                if missing_caps:
+                    _LOGGER.warning(
+                        "Missing required Backblaze B2 capabilities for Key ID '%s': %s",
+                        user_input[CONF_KEY_ID],
+                        ", ".join(sorted(missing_caps)),
+                    )
+                    errors["base"] = "invalid_capability"
+                    placeholders["missing_capabilities"] = ", ".join(
+                        sorted(missing_caps)
+                    )
+
+            configured_prefix: str = user_input[CONF_PREFIX]
+            allowed_prefix = allowed.get("namePrefix") or ""
+            # Ensure configured prefix starts with Backblaze B2's allowed prefix
+            if allowed_prefix and not configured_prefix.startswith(allowed_prefix):
+                errors[CONF_PREFIX] = "invalid_prefix"
+                placeholders["allowed_prefix"] = allowed_prefix
```
Please move the lines which don't raise to an `else`-clause

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,614 @@
+            }
+        ).encode("utf-8")
+
+        if backup.size > MAX_BACKUP_SIZE:
+            raise BackupAgentError(
+                f"Backup size ({backup.size} bytes) exceeds maximum allowed size "
+                f"({MAX_BACKUP_SIZE} bytes)"
+            )
```
Why does this integration need to set a maximum allowed size?

ElCruncharino replied:
Was just a thought to anticipate runaway backup sizes, in the case of some sort of a leak or misconfiguration. I have removed it.

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,614 @@
+                "Metadata file upload finished for %s", prefixed_metadata_filename
+            )
+
+        except B2Error as err:
+            _LOGGER.error("Backblaze B2 API error during backup upload: %s", err)
+            await self._cleanup_failed_upload(prefixed_tar_filename)
+            raise BackupAgentError(
+                f"Failed to upload backup to Backblaze B2: {err}"
+            ) from err
```
How come we need this exception handling when the method is wrapped in `handle_b2_errors`? Can we do the cleanup in a finally-clause instead?

ElCruncharino replied:
Sure. That works too. Makes sense to me.

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,614 @@
+        Called within a thread pool executor.
+        """
+        try:
+            download_response = file_version.download().response
+            try:
+                metadata_content = _parse_metadata(
+                    download_response.content.decode("utf-8")
+                )
+            except ValueError:
+                return None, None
+
+            if metadata_content["backup_id"] == target_backup_id:
+                found_backup_file = _find_backup_file_for_metadata(
+                    file_name, all_files_in_prefix, self._prefix
+                )
+                if found_backup_file:
+                    _LOGGER.debug(
+                        "Found backup file %s and metadata file %s for ID %s",
+                        found_backup_file.file_name,
+                        file_name,
+                        target_backup_id,
+                    )
+                    return found_backup_file, file_version
+                _LOGGER.warning(
+                    "Found metadata file %s for backup ID %s, but no corresponding backup file",
+                    file_name,
+                    target_backup_id,
+                )
+            _LOGGER.debug(
+                "Metadata file %s does not match target backup ID %s",
+                file_name,
+                target_backup_id,
+            )
+        except B2Error as err:
+            _LOGGER.warning(
+                "Failed to download metadata file %s during ID search: %s",
+                file_name,
+                err,
+            )
+        return None, None
```
We generally prefer guard clauses and early returns over nesting. Try to rewrite to something like this:
```suggestion
    def _process_metadata_file_for_id_sync(
        self,
        file_name: str,
        file_version: FileVersion,
        target_backup_id: str,
        all_files_in_prefix: dict[str, FileVersion],
    ) -> tuple[FileVersion | None, FileVersion | None]:
        """Synchronously process a single metadata file for a specific backup ID.

        Called within a thread pool executor.
        """
        try:
            download_response = file_version.download().response
        except B2Error as err:
            _LOGGER.warning(
                "Failed to download metadata file %s during ID search: %s",
                file_name,
                err,
            )
            return None, None

        try:
            metadata_content = _parse_metadata(
                download_response.content.decode("utf-8")
            )
        except ValueError:
            return None, None

        if metadata_content["backup_id"] != target_backup_id:
            _LOGGER.debug(
                "Metadata file %s does not match target backup ID %s",
                file_name,
                target_backup_id,
            )
            return None, None

        if not (found_backup_file := _find_backup_file_for_metadata(
            file_name, all_files_in_prefix, self._prefix
        )):
            _LOGGER.warning(
                "Found metadata file %s for backup ID %s, but no corresponding backup file",
                file_name,
                target_backup_id,
            )
            return None, None

        _LOGGER.debug(
            "Found backup file %s and metadata file %s for ID %s",
            found_backup_file.file_name,
            file_name,
            target_backup_id,
        )
        return found_backup_file, file_version
```

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,614 @@
+            self._all_files_cache_expiration = time() + CACHE_TTL
+            return all_files_in_prefix
+
+    def _process_metadata_file_sync(
+        self,
+        file_name: str,
+        file_version: FileVersion,
+        all_files_in_prefix: dict[str, FileVersion],
+    ) -> AgentBackup | None:
+        """Synchronously process a single metadata file and return an AgentBackup if valid."""
+        try:
+            download_response = file_version.download().response
+            try:
+                metadata_content = _parse_metadata(
+                    download_response.content.decode("utf-8")
+                )
+            except ValueError:
+                return None
+
+            found_backup_file = _find_backup_file_for_metadata(
+                file_name, all_files_in_prefix, self._prefix
+            )
+
+            if found_backup_file:
+                _LOGGER.debug(
+                    "Successfully processed metadata file %s for backup ID %s",
+                    file_name,
+                    metadata_content["backup_id"],
+                )
+                return _create_backup_from_metadata(metadata_content, found_backup_file)
+
+            _LOGGER.warning(
+                "Found metadata file %s but no corresponding backup file",
+                file_name,
+            )
+
+        except B2Error as err:
+            _LOGGER.warning("Failed to download metadata file %s: %s", file_name, err)
+        return None
```
Same comment, rewrite to guard clauses with early returns

emontnemery reviewed (changes requested):
Looks good! Some minor nits. I've not reviewed the tests yet.

ElCruncharino committed:
Update homeassistant/components/backblaze_b2/backup.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
rewrite guard clauses for _process_metadata_file_for_id_sync

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
AsyncIteratorReader: export or backup util. awaiting PR

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
refactor some ugly functions, address pull request feedback

ElCruncharino committed:
refactor backblaze test suite for maintainability

reduce test file from 808 to 649 lines, add agent fixture to eliminate repetitive setup, consolidate similar error tests using parametrize, simplify test patterns and remove line number references, maintain 100% coverage with all linting checks passing

ElCruncharino committed:
WIP: test simplification. Get tests more in line with other backup integrations.

ElCruncharino commented:
> Looks good! Some minor nits. I've not reviewed the tests yet.

I have refactored the tests considerably. So hopefully they're a little easier for you to process.

ElCruncharino committed:
Update to try block for streaming

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
Except and else order change for partial cleanup of main backup file

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
Per PR #153194, AsyncIteratorReader was moved away from backup integration

Co-authored-by: Erik Montnemery <erik@montnemery.com>

ElCruncharino committed:
Addresses recent Pull Request changes requested

ElCruncharino committed:
Add backwards compatibility for AsyncIteratorReader in backblaze backup

AsyncIteratorReader was moved from backup.util to util.async_iterator
in PR #153194. Add fallback import to support both locations.

ElCruncharino committed:
Handle AsyncIteratorReader signature change in backblaze backup

The signature changed from (hass, stream) to (loop, stream) in PR #153194.
Add runtime check to use the correct signature based on which version is available.

emontnemery commented on tests/components/backblaze_b2/conftest.py:
```diff
@@ -0,0 +1,368 @@
+            if original_upload_local_file is not None:
+                b2sdk.v2.bucket.Bucket.upload_local_file = original_upload_local_file
+            else:
+                delattr(b2sdk.v2.bucket.Bucket, "upload_local_file")
```
What's going on here, why are we monkey patching the library instead of using `unittest.patch`?

emontnemery reviewed (changes requested):
The implementation looks good now! 🎉 

A couple of initial comments on the tests.

emontnemery reviewed (changes requested):
Please try to rewrite the tests in `test_backup.py` according to the comment

ElCruncharino committed:
added semaphores and test to prevent urllib3 connection pool exhaustion

ElCruncharino commented:
> @ElCruncharino ,
> 
> Seeing the following warnings now within my log file:
> 
> ```
> 2025-10-25 12:48:09.060 WARNING (SyncWorker_1) [urllib3.connectionpool] Connection pool is full, discarding connection: f000.backblazeb2.com. Connection pool size: 10
> ```

My guess is you have more than 10 backups. urllib3, I think defaults to a connection pool size of 10 (as the error shows), and its getting exhausted because my component tries to download all the backup metadata files at once. As a fix of sorts, I added a semaphore that limits the concurrent metadata downloads to 5 at a time. Perhaps this will fix it? Let me know if you still see the warning.

It should leave enough headroom to process the backups efficients and not generate that warning. 

@emontnemery, open to suggestions on if there's a better way to handle this. 

I also added some tests for this scenario, as I honestly hadn't thought about it. So thanks @ehendrix23 for the feedback!

ehendrix23 commented:
> > @ElCruncharino ,
> > Seeing the following warnings now within my log file:
> > ```
> > 2025-10-25 12:48:09.060 WARNING (SyncWorker_1) [urllib3.connectionpool] Connection pool is full, discarding connection: f000.backblazeb2.com. Connection pool size: 10
> > ```
> 
> My guess is you have more than 10 backups. urllib3, I think defaults to a connection pool size of 10 (as the error shows), and its getting exhausted because my component tries to download all the backup metadata files at once. As a fix of sorts, I added a semaphore that limits the concurrent metadata downloads to 5 at a time. Perhaps this will fix it? Let me know if you still see the warning.
> 
> It should leave enough headroom to process the backups efficients and not generate that warning.
> 
> @emontnemery, open to suggestions on if there's a better way to handle this.
> 
> I also added some tests for this scenario, as I honestly hadn't thought about it. So thanks @ehendrix23 for the feedback!

No problem. 

Do you have a temporary update one available also for 2025.10 as the PR version doesn’t work on 2025.10. 😀

Thank you for doing this. Been backing up now to Backblaze for Home Assistant. Before I had to limit what to back up due to size limit. Now I just do everything. 

So thank you!!!!

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,648 @@
+        self._all_files_cache_lock = asyncio.Lock()
+        self._backup_list_cache_lock = asyncio.Lock()
+        # Limit concurrent metadata downloads to avoid exhausting connection pool
+        self._download_semaphore = asyncio.Semaphore(5)
```
The magic number 5 was chosen because of some default in urrlib.
Can we make it a constant, and then add a test which fails if whatever we rely on in urrlib is changed?
Or, can we pass some parameter to urrlib instead of relying on the default?

ElCruncharino replied:
Done added tests too.

ElCruncharino committed:
Address review concerns for semaphore and testing

- Made magic number 5 a constant (MAX_CONCURRENT_DOWNLOADS)
- Added test_requests_pool_maxsize_assumption() to verify urllib assumptions
- Parametrized test_semaphore_limits_concurrent_metadata_downloads to test both proper limiting and detection of pool exhaustion
- Updated docstrings for 6 tests to explain why they need direct agent access

ElCruncharino commented on tests/components/backblaze_b2/test_backup.py:
```diff
@@ -0,0 +1,828 @@
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test exception handling for nonexistent backup."""
```
Updated. If you can think of different ways to go about those tests, open to suggestions. But those were the walls I ran into

ElCruncharino commented on tests/components/backblaze_b2/test_backup.py:
```diff
@@ -0,0 +1,828 @@
+    assert "test_file_id" in caplog.text
+
+
+async def test_semaphore_limits_concurrent_metadata_downloads(
```
Is this feasible with mocked HTTP calls like the download? Not sure if I can check for those warnings that way. urllib3 never gets involved with `file.download()` so there won't be any warnings.

I parametrized it as you requested. The second case in my code is `(100, 15)` and basically detects the concurrent operations themselves (15 concurrent with semaphore of 100 which would exhaust the urllib3 default of 10). So it shows the test can detect when the semaphore isn't limiting.

Instead of checking logs for warnings, it's directly tracking concurrent operations with threading. I think this actually measures the concurrency we care about. This felt like the closest I could get to your suggestion without real HTTP calls. 

Does this work or should I try a different approach?

ElCruncharino commented on tests/components/backblaze_b2/test_backup.py:
```diff
@@ -0,0 +1,914 @@
+    """Test deleting backup."""
+    client = await hass_ws_client(hass)
+    with patch(
+        "homeassistant.components.backblaze_b2.backup.BackblazeBackupAgent._find_file_and_metadata_version_by_id"
```
Take a look at what I did in my latest commit (I also discovered what I believe was a small dead code path).

Your input was really helpful here to migrate to the BucketSimlulator (which admittedly I should've invested in adapting to my tests much sooner, but I struggled with it when I first looked at it a few weeks ago). 

At this point, all the patching is on a class level or via this simulator and we use the HTTP/WebSocket APIs for the tests. No imports of BackblazeBackupAgent or monkeying with it.

gjohansson-ST commented:
Please rebase this from `dev`

ElCruncharino committed:
eliminated monkey patching and unused parameter linting adjustments

ElCruncharino committed:
Merge branch 'dev' into integration-backblaze-b2

emontnemery commented on homeassistant/components/backblaze_b2/backup.py:
```diff
@@ -0,0 +1,651 @@
+            _LOGGER.exception("An error occurred during upload for %s:", filename)
+            raise
+        else:
+            _LOGGER.info(
```
We changed the recommendation about this some time ago:
- The default log level is `warning`, which means we only inform users about warning and errors in the logs
- `info` and `debug` are both allowed for integrations to use for non user facing messages

ElCruncharino committed:
Merge branch 'dev' into integration-backblaze-b2

ElCruncharino committed:
address reviewer changes, mostly to logging behavior (removed semaphore)

ElCruncharino commented on homeassistant/components/backblaze_b2/__init__.py:
```diff
@@ -0,0 +1,148 @@
+    try:
+        bucket = await hass.async_add_executor_job(_authorize_and_get_bucket_sync)
+    except exception.Unauthorized as err:
+        _LOGGER.error(
```
Done.

ElCruncharino commented on homeassistant/components/backblaze_b2/__init__.py:
```diff
@@ -0,0 +1,148 @@
+            translation_domain=DOMAIN,
+            translation_key="invalid_auth",
+        ) from err
+    except Exception as err:
```
Understood. Removed the extra logger stuff too, per your request above.

MartinHjelmare reviewed (approved):
Looks good!

emontnemery reviewed (approved):
Very nice @ElCruncharino 👍

joostlek commented:
@hugo-vrijswijk and @ElCruncharino Can you send me a message on Discord?
